packages/flutter_tools/bin/macos_assemble.sh

```diff
@@ -17,85 +17,124 @@ EchoError() {
+  # if it doesn't already exist).
+  local xcode_frameworks_dir="${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}"
+  RunCommand mkdir -p -- "${xcode_frameworks_dir}"
+  RunCommand rsync -av --delete "${ephemeral_dir}/App.framework" "${xcode_frameworks_dir}"
 
-RunCommand "${FLUTTER_ROOT}/bin/flutter"                                    \
-    ${verbose_flag}                                                         \
-    ${flutter_engine_flag}                                                  \
-    ${local_engine_flag}                                                    \
-    assemble                                                                \
-    ${performance_measurement_option}                                       \
-    -dTargetPlatform=darwin-x64                                             \
-    -dTargetFile="${target_path}"                                           \
-    -dBuildMode="${build_mode}"                                             \
-    -dTreeShakeIcons="${TREE_SHAKE_ICONS}"                                  \
-    -dDartObfuscation="${DART_OBFUSCATION}"                                 \
-    -dSplitDebugInfo="${SPLIT_DEBUG_INFO}"                                  \
-    -dTrackWidgetCreation="${TRACK_WIDGET_CREATION}"                        \
-    ${bundle_sksl_path}                                                     \
-    ${code_size_directory}                                                  \
-    --DartDefines="${DART_DEFINES}"                                         \
-    --ExtraGenSnapshotOptions="${EXTRA_GEN_SNAPSHOT_OPTIONS}"               \
-    --ExtraFrontEndOptions="${EXTRA_FRONT_END_OPTIONS}"                     \
-    --build-inputs="${build_inputs_path}"                                   \
-    --build-outputs="${build_outputs_path}"                                 \
-    --output="${ephemeral_dir}"                                             \
-   "${build_mode}_macos_bundle_flutter_assets"
+  # Embed the actual FlutterMacOS.framework that the Flutter app expects to run against,
+  # which could be a local build or an arch/type specific build.
+
+  # Copy Xcode behavior and don't copy over headers or modules.
+  RunCommand rsync -av --delete --filter "- .DS_Store/" --filter "- Headers/" --filter "- Modules/" "${ephemeral_dir}/FlutterMacOS.framework" "${xcode_frameworks_dir}/"
+
+  # Sign the binaries we moved.
+  if [[ -n "${EXPANDED_CODE_SIGN_IDENTITY:-}" ]]; then
+    RunCommand codesign --force --verbose --sign "${EXPANDED_CODE_SIGN_IDENTITY}" -- "${xcode_frameworks_dir}/App.framework/App"
+    RunCommand codesign --force --verbose --sign "${EXPANDED_CODE_SIGN_IDENTITY}" -- "${xcode_frameworks_dir}/FlutterMacOS.framework/FlutterMacOS"
+  fi
+}
+
+# Main entry point.
+if [[ $# == 0 ]]; then
+  # Unnamed entry point defaults to build.
+  BuildApp
+else
+  case $1 in
+    "build")
+      BuildApp ;;
+    "embed")
+      EmbedFrameworks ;;
+  esac
+fi
```

Comment:
This is non-trivial new bash code; any chance we could use this as the impetus to finally switch macOS over to tool_backend.* so that this could all be Dart code instead?

Reply:
I'm converging macOS and iOS scripts, and I'd like to do https://github.com/flutter/flutter/pull/70224/files#diff-cf570f69a008f028e95cfad7690ee32b3acbbc88379d340ea08ef67c0a3aff69R297 for macOS next (copy the FlutterMacOS framework directly to `BUILT_PRODUCTS_DIR` instead of `ephemeral`).

I'd rather get them to be the same bash script first since I know the iOS variant works, then swap them both over to the same dart code.  That way each PR is a good improvement that isn't doing too much and is easy to review, and if it causes issues won't be a big painful revert.

Reply:
Actually, a macOS version of https://github.com/flutter/flutter/pull/59044 needs to happen first because the `BUILT_PRODUCTS_DIR` update requires a Podfile change, but you get the idea.